### PR TITLE
Adding GiST Support for ORCA

### PR DESCRIPTION
**Pull request by ashuka24**
Prior to this PR, there was no support for GiST indexes in GPORCA.
For queries involving GiST indexes, ORCA was selecting Table Scan paths
as the optimal plan. These plans could take up to 300+ times longer than
Planner, which generated a index scan plan using the GiST index.

Example:
```
CREATE TABLE gist_tbl (a int, p polygon);
CREATE TABLE gist_tbl2 (b int, p polygon);
CREATE INDEX poly_index ON gist_tbl USING gist(p);

INSERT INTO gist_tbl SELECT i, polygon(box(point(i, i+2),point(i+4,
i+6))) FROM generate_series(1,50000)i;
INSERT INTO gist_tbl2 SELECT i, polygon(box(point(i+1, i+3),point(i+5,
i+7))) FROM generate_series(1,50000)i;

ANALYZE;
```
With the query `SELECT count(*) FROM gist_tbl, gist_tbl2 WHERE
gist_tbl.p <@ gist_tbl2.p;`, we see a performance increase with the
support of GiST.

Before:
```
EXPLAIN SELECT count(*) FROM gist_tbl, gist_tbl2 WHERE gist_tbl.p <@ gist_tbl2.p;
                                                     QUERY PLAN
---------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=0.00..171401912.12 rows=1 width=8)
   ->  Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..171401912.12 rows=1 width=8)
         ->  Aggregate  (cost=0.00..171401912.12 rows=1 width=8)
               ->  Nested Loop  (cost=0.00..171401912.12 rows=335499869 width=1)
                     Join Filter: gist_tbl.p <@ gist_tbl2.p
                     ->  Table Scan on gist_tbl2  (cost=0.00..432.25 rows=16776 width=101)
                     ->  Materialize  (cost=0.00..530.81 rows=49997 width=101)
                           ->  Broadcast Motion 3:3  (slice1; segments: 3)  (cost=0.00..525.76 rows=49997 width=101)
                                 ->  Table Scan on gist_tbl  (cost=0.00..432.24 rows=16666 width=101)
 Optimizer status: PQO version 2.65.1
(10 rows)

Time: 170.172 ms
SELECT count(*) FROM gist_tbl, gist_tbl2 WHERE gist_tbl.p <@ gist_tbl2.p;
 count
-------
 49999
(1 row)

Time: 546028.227 ms
```

After:
```
EXPLAIN SELECT count(*) FROM gist_tbl, gist_tbl2 WHERE gist_tbl.p <@ gist_tbl2.p;
                                                  QUERY PLAN
---------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=0.00..21749053.24 rows=1 width=8)
   ->  Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..21749053.24 rows=1 width=8)
         ->  Aggregate  (cost=0.00..21749053.24 rows=1 width=8)
               ->  Nested Loop  (cost=0.00..21749053.24 rows=335499869 width=1)
                     Join Filter: true
                     ->  Broadcast Motion 3:3  (slice1; segments: 3)  (cost=0.00..526.39 rows=50328 width=101)
                           ->  Table Scan on gist_tbl2  (cost=0.00..432.25 rows=16776 width=101)
                     ->  Bitmap Table Scan on gist_tbl  (cost=0.00..21746725.48 rows=6667 width=1)
                           Recheck Cond: gist_tbl.p <@ gist_tbl2.p
                           ->  Bitmap Index Scan on poly_index  (cost=0.00..0.00 rows=0 width=0)
                                 Index Cond: gist_tbl.p <@ gist_tbl2.p
 Optimizer status: PQO version 2.65.1
(12 rows)

Time: 617.489 ms

SELECT count(*) FROM gist_tbl, gist_tbl2 WHERE gist_tbl.p <@ gist_tbl2.p;
 count
-------
 49999
(1 row)

Time: 7779.198 ms
```

GiST support was implemented by sending over GiST index information to
GPORCA in the metadata using a new index enum specifically for GiST. 
GPORCA then allows GiST indexes to take both the Bitmap or B-Tree path
when creating alternatives. GPORCA also supports GiST for AOCO tables, 
but will only allow the Bitmap path for them due to the fact that B-Trees
themselves are not supported for AOCO tables by the Executor. 

This will be ported to GPDB master once we have merged https://github.com/greenplum-db/gpdb/pull/5320

Corresponding GPORCA PR: https://github.com/greenplum-db/gporca/pull/386

Signed-off-by: Bhuvnesh Chaudhary <bchaudhary@pivotal.io>
Signed-off-by: Ekta Khanna <ekhanna@pivotal.io>

### CODE REVIEW

File: src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp [hunk 1]
```diff
@@ -1374,16 +1380,22 @@ CTranslatorRelcacheToDXL::PmdindexPartTable
...
+
 	IMDIndex::EmdindexType emdindt = IMDIndex::EmdindBtree;
 	IMDId *pmdidItemType = NULL;
 	if (INDTYPE_BITMAP == pidxinfo->indType)
 	{
 		emdindt = IMDIndex::EmdindBitmap;
 		pmdidItemType = GPOS_NEW(pmp) CMDIdGPDB(GPDB_ANY);
 	}
+
+	if (INDTYPE_GIST == pidxinfo->indType)
```
Comment (oarap) [inline src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp]: `else if`

File: src/test/regress/input/qp_gist_indexes2.source [hunk 2]
```diff
@@ -104,6 +105,7 @@ INSERT INTO GistTable1 (id, owner, description, property, poli, bullseye)
 -- ----------------------------------------------------------------------------
 
 SET enable_seqscan = FALSE;
+set optimizer_enable_tablescan = FALSE;
```
Comment (dhanashreek89) [inline src/test/regress/input/qp_gist_indexes2.source]: we should trace any fallback here using `set optimizer_trace_fallback = on`, since these are not `EXPLAIN` tests and we are turning table scan off.

File: src/backend/cdb/cdbpartindex.c [hunk 3]
```diff
@@ -588,6 +588,11 @@ recordIndexesOnLeafPart(PartitionIndexNode **pNodePtr,
 			}
 		}
 
+		if (GIST_AM_OID == indRel->rd_rel->relam)
```
Comment (hardikar) [inline src/backend/cdb/cdbpartindex.c]: Can we have GiST indexes on AO tables? Otherwise let's move this in the previous if case?

File: src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp [hunk 4]
```diff
@@ -1476,7 +1476,6 @@ CTranslatorDXLToPlStmt::TranslateIndexConditions
 		OID oidOpFamily = CTranslatorUtils::OidIndexQualOpFamily(iAttno, CMDIdGPDB::PmdidConvert(pmdindex->Pmdid())->OidObjectId());
 		GPOS_ASSERT(InvalidOid != oidOpFamily);
 		gpdb::IndexOpProperties(oidCmpOperator, oidOpFamily, &iSN, &oidIndexSubtype, &fRecheck);
-		GPOS_ASSERT(!fRecheck);
```
Comment (hardikar) [inline src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp]: Why is this removed?

File: src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp [hunk 5]
```diff
@@ -1118,7 +1118,13 @@ CTranslatorRelcacheToDXL::Pmdindex
 			emdindt = IMDIndex::EmdindBitmap;
 			pmdidItemType = GPOS_NEW(pmp) CMDIdGPDB(GPDB_ANY);
 		}
-		
+
+		if (GIST_AM_OID == relIndex->rd_rel->relam)
```
Comment (hardikar) [inline src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp]: else if

File: src/backend/cdb/cdbpartindex.c [hunk 6]
```diff
@@ -1503,6 +1508,11 @@ logicalIndexInfoForIndexOid(Oid rootOid, Oid indexOid)
 		}
 	}
 
+	if (GIST_AM_OID == indRel->rd_rel->relam)
```
Comment (hardikar) [inline src/backend/cdb/cdbpartindex.c]: Here also.

File: src/test/regress/input/qp_gist_indexes2.source (hunk 2 above)
Comment (ashuka24) [reply src/test/regress/input/qp_gist_indexes2.source]: we are addressing this by removing the ignores around the explain statements.

File: src/backend/cdb/cdbpartindex.c (hunk 3 above)
Comment (ashuka24) [reply src/backend/cdb/cdbpartindex.c]: Yes, we can have GiST indexes on AO tables!

File: src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp (hunk 4 above)
Comment (ashuka24) [reply src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp]: For index scans on GiST, we want an additional recheck condition. Also, `fRecheck` is only true of `amopreqcheck` in `pg_amop` is true. This column is no longer part of the catalog after PostgreSQL 8.4

File: src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp (hunk 5 above)
Comment (ashuka24) [reply src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp]: We cannot have else if here because GiST does support AO tables

File: src/backend/cdb/cdbpartindex.c (hunk 6 above)
Comment (ashuka24) [reply src/backend/cdb/cdbpartindex.c]: Same reasoning here, we support AO for GiST so we don't want the `plogicalIndexInfo->indType` to be Bitmap

File: src/backend/cdb/cdbpartindex.c [hunk 7]
```diff
@@ -577,15 +577,14 @@ recordIndexesOnLeafPart(PartitionIndexNode **pNodePtr,
...
 		{
-			if (BTREE_AM_OID == indRel->rd_rel->relam)
-			{
-				indType = INDTYPE_BTREE;
-			}
+			indType = INDTYPE_GIST;
+		}
+		else if (RELSTORAGE_HEAP == relstorage && BTREE_AM_OID == indRel->rd_rel->relam)
+		{
+			/* for AO and AOCO tables we assume a b-tree index is bitmap else we send it as B-Tree to ORCA */
```
Comment (hardikar) [inline src/backend/cdb/cdbpartindex.c]: This comment is confusing inside an if condition for HEAP tables. Because now we can no longer assume AO/AOCO have a bitmap index (since there is a GIST index check above).

File: src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp (hunk 5 above)
Comment (hardikar) [reply src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp]: I don't like the logic here then. If its AOCO or AO first set emdindt to Bitmap. Then check if its GIST and override it afterwards.

File: src/test/regress/expected/qp_gist_indexes3_optimizer.out [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+-- start_ignore
+create schema qp_gist_indexes3;
+set search_path to qp_gist_indexes3;
+-- end_ignore
+-- ----------------------------------------------------------------------
+-- Test: test01CreateTable.sql
+-- ----------------------------------------------------------------------
+------------------------------------------------------------------------------
+-- start_ignore
+-- Portions Copyright (c) 2010, Greenplum, Inc.  All rights reserved.
```
Comment (bhuvnesh2703) [inline src/test/regress/expected/qp_gist_indexes3_optimizer.out]: do we need licensing here. i think no (or only Pivotal Software line is good enough)
